### PR TITLE
Refactor storage classes and update testcases

### LINKED ISSUE #57: Add auto save to volunteers and elderly
Currently, only person objects are autosaved and stored in the `friendlylink.json` file

TODOs:
- Add auto save ability to volunteer and elderly commands
- Currently, the autosave behaviour is executed in the Logic Manager after each command is executed. This is desirable only if everything is saved in the same file. However, since each entity is saved to their own file, lets refactor this into something more desirable.

### LINKED ISSUE #58: Make `FriendlyLinkStorage`, `VolunteerStorage` and `ElderlyStorage` extend from Storage instead of the other way round

### LINKED ISSUE #59: Add test cases for the Volunteer Storage

### LINKED ISSUE #60: Add test cases for Elderly Storage

### LINKED ISSUE #62: Make `JsonSerializable<Entity>` extends from `JsonSerializable`

### PR DESCRIPTION
**Pull request by Zxun2**
Fixes #62, Fixes #59, Fixes #60, Fixes #58, Fixes #57

## Changes
- Added more abstractions for storage classes. These classes now extend from `JsonAppStorage` which provides general methods for reading and retrieving data.
- Update behaviour for `initModelManager` to read all relevant JSON files and store them in the application cache.
- `JsonAdaptedVolunteer` and `JsonAdaptedElderly` now extend from `JsonAdaptedPerson`
- Serializable entities now implement the `JsonSerializable` interface which ensures they override the `toModelType` method
- Add test cases for Volunteer and Elderly Storage
- Overall improvement in code quality from the previous sprint.  

### Notes

To clarify, the current storage structure is as follows: 
- `FriendlyLinkStorage` -- responsible for application state (i.e. summary statistics if we were to implement them)
- `ElderlyStorage` -- for Elderly
- `VolunteerStorage` -- for Volunteer 

On the other hand, we have `FriendlyLink` implements `ReadOnlyFriendlyLink` which extends from `ReadOnlyElderly` and `ReadOnlyVolunteer`. The `FriendlyLink` class refers to the application-wide cache that holds the `UniqueList` for all entities.

### TODO

- [x] implement quick-hax autosave for now

### CODE REVIEW

**Comment by wz2k** [comment]
hi

**Review by gohyongjing — APPROVED** [review]
LGTM! Perhaps we can merge this first and i can amend the structure and you can take a look if my PR is okay.

File: src/main/java/seedu/address/MainApp.java [hunk 1]
```diff
@@ -82,28 +79,21 @@ private void initializeAppManagers(
...
-     * Returns a {@code ModelManager} with the data from {@code storage}'s address book and {@code userPrefs}. <br>
-     * The data from the sample address book will be used instead if {@code storage}'s address book is not found,
-     * or an empty address book will be used instead if errors occur when reading {@code storage}'s address book.
+     * Returns a {@code ModelManager} with the data from {@code AppStorage} and {@code userPrefs}. <br>
+     * An empty application will be used instead if errors occur when reading {@code storage}.
      */
     private Model initModelManager(Storage storage, ReadOnlyUserPrefs userPrefs) {
-        Optional<ReadOnlyFriendlyLink> friendlyLinkOptional;
-        ReadOnlyFriendlyLink initialData;
+        FriendlyLink applicationCache = new FriendlyLink();
```
[the remark this answers is not in the document]
Comment (Zxun2) [reply src/main/java/seedu/address/MainApp.java:86]: Yup, agreed that some optimizations can be made here.

[the remark this answers is not in the document]
Comment (gohyongjing) [reply src/main/java/seedu/address/MainApp.java:86]: Lets change that in another pull request

**Pull request merged**